coreTable/CoreTable.ts

```diff
@@ -1191,11 +1191,85 @@ export class CoreTable<
 
     fullJoin(rightTable: CoreTable, by?: ColumnSlug[]): this {
         return this.leftJoin(rightTable, by)
-            .concat(rightTable.leftJoin(this, by))
+            .concat([rightTable.leftJoin(this, by)])
             .dropDuplicateRows()
     }
 
-    groupBy(by: ColumnSlug[]) {
-        // todo
+    indexBy(slug: ColumnSlug) {
+        const map = new Map<CoreValueType, number[]>()
+        this.getValuesFor(slug).map((value, index) => {
+            if (!map.has(value)) map.set(value, [])
+            map.get(value)!.push(index)
+        })
+        return map
+    }
+
+    groupBy(by: ColumnSlug) {
+        const index = this.indexBy(by)
+        return Array.from(index.keys()).map((groupName) =>
+            this.transform(
+                this.columnStore,
+                this.defs,
+                `Rows for group ${groupName}`,
+                TransformType.FilterRows,
+                new FilterMask(this.numRows, index.get(groupName)!)
+            )
+        )
+    }
+
+    reduce(reductionMap: ReductionMap) {
+        const lastRow = { ...this.lastRow }
+        Object.keys(reductionMap).forEach((slug) => {
+            const prop = reductionMap[slug]
+            const col = this.get(slug)!
+            if (typeof prop === "string") lastRow[slug] = col[prop]
+            else lastRow[slug] = prop(col)
+        })
+        return this.transform(
+            rowsToColumnStore([lastRow]),
+            this.defs,
+            `Reduced table`,
+            TransformType.Reduce
+        )
+    }
+}
+
+interface ReductionMap {
+    [columnSlug: string]:
+        | ReductionTypes
+        | ((column: CoreColumn) => CoreValueType)
+}
+
+type ReductionTypes = keyof CoreColumn
+
+class FilterMask {
+    private mask: boolean[]
+    private numRows: number
+    constructor(
+        numRows: number,
+        input: boolean[] | number[],
+        keepThese = true
+    ) {
+        this.numRows = numRows
+        if (typeof input[0] === "number") {
+            const set = new Set(input as number[])
+            this.mask = range(0, numRows).map((index) =>
+                set.has(index) ? keepThese : !keepThese
+            )
+        } else this.mask = input as boolean[]
+    }
+
+    inverse() {
+        return new FilterMask(this.numRows, this.mask.reverse())
+    }
```

Comment:
@breck7 Should this be `reverse()`? `inverse()` I understand as flipping the booleans, from TRUE → FALSE and vice versa.

Reply:
Ah yes! Nice catch. I'm guessing my test passes because I have only 2 samples. Will take a look.

Reply:
https://github.com/owid/owid-grapher/commit/256bcb325c6a486f5c709caf618456a4788bcecf